modules/private/class-jetpack-private.php

```diff
@@ -0,0 +1,313 @@
+<?php
+/**
+ * The Private Site module class file.
+ *
+ * @package Jetpack
+ */
+
+/**
+ * Class Jetpack_Private
+ */
+class Jetpack_Private {
+	/**
+	 * Init function that adds all hooks needed for the module.
+	 */
+	public static function init() {
+		add_action( 'parse_request', array( __CLASS__, 'privatize_blog' ), 100 );
+		add_action( 'login_init', array( __CLASS__, 'privatize_blog_maybe_mask_blog_name' ) );
+		add_filter( 'preprocess_comment', array( __CLASS__, 'privatize_blog_comments' ), 0 );
+		add_action( 'blog_privacy_selector', array( __CLASS__, 'privatize_blog_priv_selector' ) );
+		add_filter( 'robots_txt', array( __CLASS__, 'private_robots_txt' ) );
+		add_action( 'wp_head', array( __CLASS__, 'private_no_pinning' ) );
+		add_action( 'admin_init', array( __CLASS__, 'private_blog_prevent_requests' ), 9 );
+		add_action( 'check_ajax_referer', array( __CLASS__, 'private_blog_ajax_nonce_check' ), 9, 2 );
+		add_action( 'rest_pre_dispatch', array( __CLASS__, 'disable_rest_api' ) );
+		add_filter( 'jetpack_get_available_modules', array( __CLASS__, 'private_get_modules' ) );
```

Comment:
I know you've discussed this a bit in this PR. Do you think you could clarify the expected behaviour a bit in the PR description / testing instructions?

Right now and since you rely on `jetpack_get_available_modules` only, some modules are hidden from the Jetpack dashboard: you cannot turn them on manually. However, since you do not use the `jetpack_get_default_modules` filter, there are no changes to the list of modules that get automatically activated when you connect a site to your WordPress.com, or to the modules that remain active after you turn on the Private module. As a result, you have a few modules that get automatically activated, such as Sharing, Subscriptions, Publicize, or the  JSON API, or that you may have activated in the past before the Private Sites feature was turned on, but those features do not appear in the Jetpack dashboard anymore so:
1. You don't know they are activated.
2. You can't deactivate them.

This is a problem with some of those features, because you can stumble across them accidentally, start using them (assuming your site is private), and share links to your site publicly. As an example, you'll find links to the Publicize settings screen when writing posts:

<img width="298" alt="screenshot 2019-04-23 at 19 08 28" src="https://user-images.githubusercontent.com/426388/56609824-5699b500-660e-11e9-8bcd-4be331107615.png">

I realize you will change the behaviour a bit for Publicize, but I don't know if that will be very obvious for site owners.

In addition to that, other features are impacted. You can for example also find a Settings > Sharing link in wp-admin, leading you to the Sharing buttons options screen.

What is the expected behaviour here? When I turn on the feature, should all those modules stop working / be deactivated? In this case, this may be important:
https://jetpack.com/support/module-overrides/

Reply:
The intended behaviour is that when this module is activated, all of the modules in `private_get_modules` are deactivated. Then when this module is deactivated, these modules are reactivated. 

The reason for using `jetpack_get_available_modules` rather than `jetpack_active_modules`, is that the settings for `jetpack_active_modules` stick even when they module is deactivated. Does that make sense? Do you have a different suggestion for how we could handle this?

Reply:
> The intended behaviour is that when this module is activated, all of the modules in private_get_modules are deactivated.

Right, this makes a lot of sense. Unfortunately that is not what is happening right now. 

See p1556092075145300-slack-jetpack-crew for more discussion about this.